### DISCUSSION TITLE
inherit path from function name instead of random _data/<id>

### DISCUSSION THREAD

**Discussion by sharmapukar217** [rakkasjs/rakkasjs#70] [Ideas]
When using useSSQ and useSSM, it seems that the url seems to be generated like `_data/<some_id>`. It is possible to generate the path from the function name. for eg useSSQ(async function getCurrentUser() {
 // stuffs
}), would be generating a fetch request to `/_data/getCurrentUser` or maybe more flexible by adding `name` or `path` name the the config passed to useSSQ as 2nd parameter. This allow one to debug the requests by inspecting the network tab.

**Comment by cyco130** [comment]
I've been thinking about this a lot, it is a real problem that needs to be solved. So far I've been leaning towards the second parameter approach. Here's some documentation on how it works currently:

Currently, the `_data` URLs have the form `_data/<BUILD_ID>/<MODULE_ID>/<INDEX_NO>/<CLOSURE>` where:

`BUILD_ID` is a build ID, derived from git hash by default. In development, it's always `development` though. You can override it by setting the `RAKKAS_BUILD_ID` environment variable. It's there to protect against URL clashes between deploys. I'm thinking of adding functionality to do a full reload in case of a request with a stale build ID.

`MODULE_ID` is a random module ID. During development, it's the full path to the module, it has to be that way because Vite doesn't analyze the full bundle so it's the easiest way to guarantee a one-on-one mapping between modules and IDs.

`INDEX_NO` is simply the index number of the `useSSQ` (or friends) call in that particular module. If you have only one call, it's `0`, if you have two, the second one will be `1` etc.

`CLOSURE` is only available when a `GET` request is made which is the default for `run/useServerSideQuery` (but not `run/useServerSideMutation`). It's sent in the body for `POST` requests. It contains the captured values from the surrounding scope. The reason for using a `GET` request is to make it possible to prerender statically (actually there's a known bug here: The encoding currently used by Rakkas is case sensitive, which may cause collisions on Windows). Another reason will be to enable caching: Currently there's no way to set the response headers but it can be added easily and then you can set the `Cache-Control` header to enable HTTP caching.

When we implement custom IDs, we can replace the `MODULE_ID`-`INDEX_NO` pair with a user-defined ID and do a lookup from a manifest file created during the build. It may have to stay more verbose for dev but I don't think it'll be a big problem.

We can even designate some calls as "stable" so that the build ID isn't used for that particular call and it's possible to use the same URL across deploys, just like a normal API route.
